### PR TITLE
Kionix KX022 3-axis accelerometer driver

### PR DESCRIPTION
**Pull request by AzamLukman**
Adding support Kionix KX022 3-axis accelerometer driver.

Product link:  https://www.kionix.com/product/KX022-1020

### CODE REVIEW

**Review by AzamLukman — COMMENTED** [review]
done checking

File: drivers/sensor/kx022/Kconfig [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+	int "Resolution of the KX022"
+	range 0 2
+	default 2
+	help
+	  AKA Sensitivity in the datasheet, related to the power of the
+	  KX022, the default setting is 16-bit resolution.
+	  0: RES selected at runtime
+	  1: 8-bit
+	  2: 16-bit
+endmenu
```
Comment (gmarull) [inline drivers/sensor/kx022/Kconfig:186]: please move all these configurations to Devicetree bindings

File: drivers/sensor/kx022/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_library()
+
+zephyr_library_sources(kx022.c)
+zephyr_library_sources(kx022_i2c.c)
```
Comment (gmarull) [inline drivers/sensor/kx022/CMakeLists.txt:6]: `zephyr_library_sources(kx022.c kx022_i2c.c)`

File: drivers/sensor/kx022/kx022.c [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+#define kx022_delay(msec) k_msleep(((uint16_t)msec) * 1000 / CONFIG_KX022_ODR + 1);
+
+#if defined(KX022_ODR_RUNTIME)
+static const uint16_t kx022_hr_odr_map[] = { 12, 25, 50, 100, 200, 400, 800, 1600 };
+
+static int kx022_freq_to_odr_val(uint16_t freq)
+{
+	size_t i;
+
+	for (i = 0; i < ARRAY_SIZE(kx022_hr_odr_map); i++) {
```
Comment (gmarull) [inline drivers/sensor/kx022/kx022.c:49]: add `U` suffix to unsigned constants

File: drivers/sensor/kx022/kx022.c [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+	if (odr < 0) {
+		return odr;
+	}
+
+	kx022_mode(KX022_STANDY_MODE);
+
+	if (data->hw_tf->update_reg(data, KX022_REG_ODCNTL, KX022_MASK_ODCNTL_OSA, (uint8_t)odr) <
+	    0) {
+		LOG_DBG("Failed to set KX022 sampling rate");
+		return -EIO;
```
Comment (gmarull) [inline drivers/sensor/kx022/kx022.c:73]: please propagate error codes,

```
ret = f();
if (ret < 0) {
    return ret;
}
```

File: drivers/sensor/kx022/kx022.c [hunk 5]
```diff
@@ -0,0 +1,523 @@
...
+	for (val = 0; val < ARRAY_SIZE(kx022_accel_res); val++) {
+		if (res == kx022_accel_res[val]) {
+			goto SET_RES;
+		}
+	}
+
+	LOG_DBG("KX022 resolution invalid");
+	return -EIO;
+
+SET_RES:
```
Comment (gmarull) [inline drivers/sensor/kx022/kx022.c:140]: goto can be avoided here, doesn't provide any benefit

File: drivers/sensor/kx022/kx022.c [hunk 6]
```diff
@@ -0,0 +1,523 @@
...
+		LOG_ERR("Invalid register address");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+#define kx022_sample_fetch_accel_x(dev) kx022_sample_fetch_accel(dev, KX022_REG_XOUT_L)
+#define kx022_sample_fetch_accel_y(dev) kx022_sample_fetch_accel(dev, KX022_REG_YOUT_L)
+#define kx022_sample_fetch_accel_z(dev) kx022_sample_fetch_accel(dev, KX022_REG_ZOUT_L)
```
Comment (gmarull) [inline drivers/sensor/kx022/kx022.c:296]: why is this needed? just use fetch directly

File: drivers/sensor/kx022/kx022.c [hunk 7]
```diff
@@ -0,0 +1,523 @@
...
+	uint8_t buf[6];
+
+	if (data->hw_tf->read_data(data, KX022_REG_XOUT_L, buf, sizeof(buf)) < 0) {
+		LOG_DBG("failed to read sample");
+		return -EIO;
+	}
+
+	data->sample_x = (int16_t)((uint16_t)(buf[0]) | ((uint16_t)(buf[1]) << 8));
+	data->sample_y = (int16_t)((uint16_t)(buf[2]) | ((uint16_t)(buf[3]) << 8));
+	data->sample_z = (int16_t)((uint16_t)(buf[4]) | ((uint16_t)(buf[5]) << 8));
```
Comment (gmarull) [inline drivers/sensor/kx022/kx022.c:310]: please use `byteorder.h` API

File: drivers/sensor/kx022/kx022.c [hunk 8]
```diff
@@ -0,0 +1,523 @@
...
+	.channel_get = kx022_channel_get,
+};
+
+static int kx022_init(const struct device *dev)
+{
+	const struct kx022_config *const config = dev->config;
+	struct kx022_data *data = dev->data;
+	uint8_t chip_id;
+
+	data->comm_master = device_get_binding(config->comm_master_dev_name);
```
Comment (gmarull) [inline drivers/sensor/kx022/kx022.c:459]: use `i2c_dt_spec` instead, see other sensor drivers

File: drivers/sensor/kx022/kx022.c [hunk 9]
```diff
@@ -0,0 +1,523 @@
...
+	if (!data->comm_master) {
+		LOG_DBG("master not found: %s", config->comm_master_dev_name);
+		return -EINVAL;
+	}
+	k_msleep(2000);
```
Comment (gmarull) [inline drivers/sensor/kx022/kx022.c:464]: - this will delay reaching `main()` with at least 2s, is this really needed?
- constants need to be defined

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/sensor/kx022/Kconfig (hunk 1 above)
Comment (ycsin) [reply drivers/sensor/kx022/Kconfig:186]: https://docs.zephyrproject.org/latest/guides/dts/bindings.html#enum-values
